src/net/connection_controller.rs

```diff
@@ -0,0 +1,158 @@
+use crate::{
+    config::Config,
+    error::Result,
+    net::{events::SocketEvent, SocketSender, VirtualConnection},
+    packet::{DeliveryGuarantee, OutgoingPackets, Packet, PacketInfo},
+};
+use crossbeam_channel::Sender;
+use log::error;
+use std::{self, net::SocketAddr, time::Instant};
+
+/// Controls all aspects of the connection:
+/// * Processes incoming data (from a socket) or events (from a user).
+/// * Updates connection state: resends dropped packets, sends heartbeat packet, etc.
+/// * Creates new connections.
+/// * Checks if connection should be dropped.
+/// It doesn't own connections, but only owns necessary components to process them.
+#[derive(Debug)]
+pub struct ConnectionController<PacketSender> {
+    config: Config,
+    packet_sender: PacketSender,
+    event_sender: Sender<SocketEvent>,
+}
+
+/// Defines a connection type.
+type Connection = VirtualConnection;
+/// Defines a user event type.
+type UserEvent = Packet;
+/// Defines a connection event type.
+type ConnectionEvent = SocketEvent;
+
+impl<PacketSender: SocketSender> ConnectionController<PacketSender> {
+    /// Creates a new instance of `ConnectionHandler`.
+    pub fn new(
+        config: Config,
+        packet_sender: PacketSender,
+        event_sender: Sender<ConnectionEvent>,
+    ) -> Self {
+        ConnectionController {
+            config,
+            packet_sender,
+            event_sender,
+        }
+    }
+
+    /// Creates new connection. Also will init it and send connection event to a user.
+    pub fn create_connection(
+        &self,
+        address: SocketAddr,
+        time: Instant,
+        initial_data: Option<&[u8]>,
+    ) -> Connection {
+        // emit connect event if this is initiated by remote host.
+        if initial_data.is_some() {
+            self.event_sender
+                .send(ConnectionEvent::Connect(address))
+                .unwrap();
+        }
+        Connection::new(address, &self.config, time)
+    }
+
+    /// Determine if this connection should be dropped due to its state.
+    pub fn should_drop(&self, connection: &mut Connection, time: Instant) -> bool {
+        let should_drop = connection.packets_in_flight() > self.config.max_packets_in_flight
+            || connection.last_heard(time) >= self.config.idle_connection_timeout;
+        if should_drop {
+            self.event_sender
+                .send(ConnectionEvent::Timeout(connection.remote_address))
+                .unwrap();
+        }
+        should_drop
+    }
+
+    /// Handle a packet received from a socket.

```

Comment:
maybe we test laminar over an in-memory transport layer, it isn't a socket always.

Reply:
Also, what do you mean with `Handle`

Reply:
By `handle` I ment something that processes, maybe my brain hasn't switched from another project, where I see lots of "handle_request" and stuff. I rename it to `process` :)

Reply:
Is this comment ok? 
`Process a received packet: parse it and emit an event.`?

Reply:
Better, it is just an comment to be specific in comments. By saying parsing and emitting an event, it makes much more sense.